### ISSUE TITLE
Implement PartialEq and Eq for all (most?) types

### ISSUE THREAD

**Issue by Pauan** [koute/stdweb#87]
Right now `Reference` uses `#[derive(PartialEq, Eq)]`, but a lot of other types don't (e.g. `Node`, `Element`, `NodeList`, etc.)

In most cases implementing it should be as simple as adding `#[derive(PartialEq, Eq)]` to the types.

Assuming you're okay with that, I'll send a pull request adding this in.

----

Also, I took a look at the implementation of `Module.STDWEB`, but I have a question. It implements `Reference` with an `i32`, and the `PartialEq` implementation simply compares the `i32`

Is it possible to have two `Reference` that have different `i32`, but they refer to the same JS object?

In other words: if `ref1 == ref2` is false in Rust, is it possible for `ref1 === ref2` to be true in JavaScript?

Because `Module.STDWEB` uses a `WeakMap` for `ref_to_id_map`, it *seems* that situation will never happen, but is it truly impossible?

**Comment by koute** [comment]
Hmmm... I *think* I might have had some reason why I didn't make everything `PartialEq` by default; alas, I don't remember at this point.

I think I would *probably* be fine with it? (I mean, right now I can't come up with any reason why not.) But maybe it might be a good idea to be conservative with it and leave it alone for now?

The question whenever it's possible to have two objects equal in JS but having two different reference IDs is indeed an interesting one. Unfortunately I don't have an answer for it. It would seem that, yes, due to how a `WeakMap` works it *should* be impossible, but JavaScript is dark and full of terrors, so there might be some way to do it which I don't know about.

**Comment by Pauan** [comment]
The reason why I am asking for this is so that I can write unit tests and examples. Being able to use `assert_eq!` with DOM nodes would be very nice.

In addition, it's expected by JavaScript developers, since JavaScript objects (and by extension DOM nodes) support equality checking.

There are plenty of JS programs that check for DOM node equality (e.g. in an event listener they check whether `event.target` is equal to a specific DOM node).

If the concern is over the correctness of the implementation of `PartialEq`, we can implement it like this:

```rust
impl PartialEq for Reference {
    #[inline]
    fn eq( &self, other: &Self ) -> bool {
        js!( return @{self.as_ref()} === @{other.as_ref()}; ).try_into().unwrap()
    }
}
```

This should ensure that it works correctly regardless of how `Reference` is implemented.
